opensrp-opd/src/main/java/org/smartregister/opd/widgets/OpdBarcodeFactory.java

```diff
@@ -0,0 +1,84 @@
+package org.smartregister.opd.widgets;
+
+import android.content.Context;
+import android.content.Intent;
+import android.support.annotation.NonNull;
+import android.view.View;
+
+import com.google.android.gms.vision.barcode.Barcode;
+import com.rengwuxian.materialedittext.MaterialEditText;
+import com.vijay.jsonwizard.constants.JsonFormConstants;
+import com.vijay.jsonwizard.fragments.JsonFormFragment;
+import com.vijay.jsonwizard.interfaces.CommonListener;
+import com.vijay.jsonwizard.interfaces.JsonApi;
+import com.vijay.jsonwizard.interfaces.OnActivityResultListener;
+import com.vijay.jsonwizard.widgets.BarcodeFactory;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.smartregister.opd.listener.LookUpTextWatcher;
+import org.smartregister.opd.utils.OpdConstants;
+
+import java.util.List;
+
+import timber.log.Timber;
+
+public class OpdBarcodeFactory extends BarcodeFactory {
+
+    private JsonFormFragment jsonFormFragment;
+    private boolean forLookUp;
+
+    @Override
+    public List<View> getViewsFromJson(@NonNull String stepName, @NonNull Context context, @NonNull JsonFormFragment formFragment,
+                                       @NonNull JSONObject jsonObject, @NonNull CommonListener listener) throws Exception {
+        List<View> viewList = super.getViewsFromJson(stepName, context, formFragment, jsonObject, listener);
+        this.jsonFormFragment = formFragment;
+        try {
+            this.forLookUp = jsonObject.has(OpdConstants.KEY.LOOK_UP) &&
+                    jsonObject.get(OpdConstants.KEY.LOOK_UP).toString().equalsIgnoreCase(Boolean.TRUE.toString());
+        } catch (JSONException e) {
+            Timber.e(e);
+        }
+        return viewList;
+    }
+
+    @Override
+    public List<View> getViewsFromJson(@NonNull String stepName, @NonNull Context context, @NonNull JsonFormFragment formFragment, @NonNull JSONObject jsonObject,
+                                       @NonNull CommonListener listener, boolean popup) {
+        List<View> viewList = super.getViewsFromJson(stepName, context, formFragment, jsonObject, listener, popup);
+        this.jsonFormFragment = formFragment;
+        try {
+            this.forLookUp = jsonObject.has(OpdConstants.KEY.LOOK_UP) &&
+                    jsonObject.get(OpdConstants.KEY.LOOK_UP).toString().equalsIgnoreCase(Boolean.TRUE.toString());
+        } catch (JSONException e) {
+            Timber.e(e);
+        }
+        return viewList;
+    }
+
+    @Override
+    protected void addOnBarCodeResultListeners(@NonNull Context context, @NonNull final MaterialEditText editText) {
+        if (context instanceof JsonApi) {
+            JsonApi jsonApi = (JsonApi) context;
+            jsonApi.addOnActivityResultListener(JsonFormConstants.BARCODE_CONSTANTS.BARCODE_REQUEST_CODE,
```

Comment:
Try and add tests for some of this code. It looks like crucial code that we depend on.

Reply:
adding the tests